### ISSUE TITLE
MapRazorComponents unexpectedly requires antiforgery middleware

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/aspnetcore#51981]
Discovered by @davidfowl. If you're trying to make a minimal Blazor Static SSR project from scratch (not from the template), you may be surprised to find that you *must* call `app.UseAntiforgery` even if you're only receiving GET requests and don't have any forms in your whole app. If not, trying to visit and `@page` component will result in the error:

![image](https://github.com/dotnet/aspnetcore/assets/1101362/191a496b-b110-4cc2-8662-5608c0f1b443)

It's a pretty weird inconsistency that you get this for `MapRazorComponents` endpoints, but do *not* get this for minimal actions that return `RazorComponentResult`.

### Possible solutions

* We could change `EndpointMiddleware` so that if it's a GET request, it doesn't try to enforce antiforgery checks. TBH I'm not sure why that's not already the case - do we even support supplying an antiforgery token with a GET request?
* Or, we could change `RazorComponentEndpointFactory` so that it registers a separate endpoint for GET vs non-GET and the GET endpoint doesn't have the `RequireAntiforgeryTokenAttribute` metadata.
* Or, we could change `MapRazorComponents` so that it somehow auto-registers the antiforgery middleware (though this is the least good solution)

**Comment by javiercn** [comment]
@SteveSandersonMS  that's a good catch.

We had some issues around in this area, but I thought we had addressed them. 

I believe what is happening is that we wire up the Antiforgery metadata on the endpoint and there's logic to enforce that you have the middleware wired up, same as for the auth middlewares. 

We do our own antiforgery handling (and respect the middleware result when present), so this should ideally not happen. We should find a way to be excluded from this check.

@captainsafia knows the details.

**Comment by captainsafia** [comment]
> We could change EndpointMiddleware so that if it's a GET request, it doesn't try to enforce antiforgery checks. TBH I'm not sure why that's not already the case - do we even support supplying an antiforgery token with a GET request?

At the moment, we only check for the metadata in the middleware and not any other properties of the request. Admittedly, this is optimized for the minimal API scenario where we only add the metadata for endpoints that absolutely require it so that no additional checks are needed on the middleware hot path.

> Or, we could change RazorComponentEndpointFactory so that it registers a separate endpoint for GET vs non-GET and the GET endpoint doesn't have the RequireAntiforgeryTokenAttribute metadata.

I'd prefer this approach. My "purist" take on this is that metadata should be the only indicator of whether you opt-in to certain behavior. It's up to the component applying the metadata (`RazorComponentEndpointFactory` or `RequestDelegateFactory`) to determine whether the metadata should be annotated.

That being said, my pragmatic side says that the first approach is valid, likely easier to implement, and has prior art [in the routing middleware](https://github.com/dotnet/aspnetcore/blob/8a539d35e7db61ce0956489f60edaa4ac0c57ce6/src/Http/Routing/src/EndpointRoutingMiddleware.cs#L173-L175).

**Comment by DeagleGross** [comment]
I want to take responsibility to close this.
TLDR: technically, this is not fixed for Antiforgery at https://github.com/dotnet/aspnetcore/pull/67082, but CSRF being the new default now mitigates the issue. 

The new template (see https://github.com/dotnet/aspnetcore/pull/67119) does not have `.UseAntiforgery()`:
```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddRazorComponents();

var app = builder.Build();

app.MapRazorComponents<App>();
app.Run();
```

`EndpointMiddleware` throws "a middleware was not found that supports anti-forgery" only when an endpoint requires antiforgery and no antiforgery-capable middleware recorded that it ran. `MapRazorComponents` adds that metadata, so previously you had to call `app.UseAntiforgery()` to satisfy the check — even for a GET-only, form-less app.

Now `ICsrfProtection` is registered by default and `CsrfProtectionMiddleware` is auto-injected into every `WebApplication`. It records the presence marker unconditionally, so the `EndpointMiddleware` check passes without any manual setup. This wasn't done by exempting GET in `EndpointMiddleware`; the default CSRF middleware is simply always wired up, so the requirement is gone for all methods. Safe methods (GET/HEAD/OPTIONS/TRACE) are allowed outright, and unsafe methods get the Fetch-Metadata cross-site check.

---

If you still meet problem with Antiforgery, and cant switch to CSRF - please let us know the reasons, and we can probably try to solve this as well.
